internal/storage/metricstore/elasticsearch/reader.go

```diff
@@ -6,47 +6,363 @@ package elasticsearch
+func (r MetricsReader) buildQuery(params metricstore.BaseQueryParameters, timeRange TimeRange) elastic.BoolQuery {
+	boolQuery := elastic.NewBoolQuery()
+
+	serviceNameQuery := elastic.NewTermsQuery("process.serviceName", buildInterfaceSlice(params.ServiceNames)...)
+	boolQuery.Filter(serviceNameQuery)
+
+	// Span kind filter
+	spanKindField := strings.ReplaceAll(model.SpanKindKey, ".", r.cfg.Tags.DotReplacement)
+	spanKindQuery := elastic.NewTermsQuery("tag."+spanKindField, buildInterfaceSlice(normalizeSpanKinds(params.SpanKinds))...)
+	boolQuery.Filter(spanKindQuery)
+
+	rangeQuery := elastic.NewRangeQuery("startTimeMillis").
+		// Use extendedStartTimeMillis to allow for a 10-minute lookback.
+		Gte(timeRange.extendedStartTimeMillis).
+		Lte(timeRange.endTimeMillis).
+		Format("epoch_millis")
+	boolQuery.Filter(rangeQuery)
+
+	// Corresponding ES query:
+	// {
+	// "query": {
+	//	"bool": {
+	//		"filter": [
+	//			{"terms": {"process.serviceName": ["name1"] }},
+	//			{"terms": {"tag.span@kind": ["server"] }}, // Dot replacement: @
+	//			{
+	//			"range": {
+	//			"startTimeMillis": {
+	//				"gte": "now-'lookback'-5m",
+	//				"lte": "now",
+	//				"format": "epoch_millis"}}}]}
+	// },
+
+	return *boolQuery
+}
+
+func getCallRateProcessMetrics(pairs []*Pair, m metricstore.BaseQueryParameters) []*Pair {
+	lookback := int(math.Ceil(float64(m.RatePer.Milliseconds()) / float64(m.Step.Milliseconds())))
+	if lookback < 1 {
+		lookback = 1 // Ensure we always have at least 1 point
+	}
+	n := len(pairs)
+	results := make([]*Pair, 0, n) // Pre-allocate result slice for efficiency
+
+	for i := range pairs {
+		// Elasticsearch's percentiles aggregation returns 0.0 for time buckets with no documents
+		// These aren't true zero values but represent missing data points in sparse time series
+		// We convert them to NaN to distinguish from actual measured zero values (slope of 0)
+		if pairs[i].Value == 0.0 {
+			results = append(results, newPair(pairs[i].TimeStamp, math.NaN()))
+			continue
+		}
+
+		// For first (lookback-1) points, we don't have enough history
+		if i < lookback-1 {
+			results = append(results, newPair(pairs[i].TimeStamp, 0.0))
+			continue
+		}
+
+		// Get boundary values for our lookback window:
+		// First value in window (oldest)
+		firstVal := pairs[i-lookback+1].Value
+		// Last value in window (current value)
+		lastVal := pairs[i].Value
+
+		// Calculate time window duration in seconds
+		// params.Step.Seconds() gives the interval between data points
+		windowSizeSeconds := float64(lookback) * (m.Step.Seconds())
+
+		// Calculate rate of change per second
+		// Formula: (current_value - starting_value) / time_window
+		rate := (lastVal - firstVal) / windowSizeSeconds
+
+		// Store the result with original timestamp
+		results = append(results, newPair(pairs[i].TimeStamp, rate))
+	}
+
+	return results
+}
+
+func getCallRateBucketsToPointsFunc(buckets []*elastic.AggregationBucketHistogramItem) []*Pair {
```

Comment:
* is it really specific to "call rate"?
* Func suffix is redundant
```suggestion
func bucketsToPoints(buckets []*elastic.AggregationBucketHistogramItem) []*Pair {
```

Reply:
Hi @yurishkuro, yes some part of it is specific to getCallRate, I want to refactor it because it's also used in GetErrorRates

Reply:
I will try to refactor this method so it's look cleaner in the future when all methods has been properly implemented

Reply:
Thanks for your feedback. Solved at [e821840](https://github.com/jaegertracing/jaeger/pull/7229/commits/e82184035f346af515400dcb46fbccedcca6c387)